hotrod/src/main/java/io/hyperfoil/hotrod/steps/HotRodRequestStep.java

```diff
@@ -0,0 +1,98 @@
+package io.hyperfoil.hotrod.steps;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+
+import io.hyperfoil.api.session.Access;
+import io.hyperfoil.api.session.ResourceUtilizer;
+import io.hyperfoil.api.session.Session;
+import io.hyperfoil.api.statistics.Statistics;
+import io.hyperfoil.core.builders.SLA;
+import io.hyperfoil.core.session.SessionFactory;
+import io.hyperfoil.core.steps.StatisticsStep;
+import io.hyperfoil.function.SerializableFunction;
+import io.hyperfoil.hotrod.api.HotRodConnectionPool;
+import io.hyperfoil.hotrod.api.HotRodDestinationTable;
+import io.hyperfoil.hotrod.api.HotRodOperation;
+import io.hyperfoil.hotrod.resource.HotRodResource;
+import io.hyperfoil.hotrod.resource.HotRodResourceKey;
+
+public class HotRodRequestStep extends StatisticsStep implements ResourceUtilizer, SLA.Provider {
+
+   final HotRodResourceKey futureWrapperKey;
+   final SerializableFunction<Session, HotRodOperation> operation;
+   final SerializableFunction<Session, String> cacheName;
+   final Access cacheKeyAccess;
+   final Access cacheValueAccess;
+
+   protected HotRodRequestStep(int id, HotRodResourceKey futureWrapperKey,
+                               SerializableFunction<Session, HotRodOperation> operation,
+                               SerializableFunction<Session, String> cacheName,
+                               Access cacheKeyAccess,
+                               Access cacheValueAccess) {
+      super(id);
+      this.futureWrapperKey = futureWrapperKey;
+      this.operation = operation;
+      this.cacheName = cacheName;
+      this.cacheKeyAccess = cacheKeyAccess;
+      this.cacheValueAccess = cacheValueAccess;
+   }
+
+   @Override
+   public SLA[] sla() {
+      return new SLA[0];
+   }
+
+   @Override
+   public boolean invoke(Session session) {
+
+      String cacheName = this.cacheName.apply(session);
+      HotRodOperation operation = this.operation.apply(session);
+
+      Object key = cacheKeyAccess.getObject(session);
+      Object value = cacheValueAccess.getObject(session);
+
+      HotRodDestinationTable destinations = HotRodDestinationTable.get(session);
+      HotRodConnectionPool pool = destinations.getPools().get(cacheName);
+      RemoteCache remoteCache = pool.getRemoteCache();
+
+      CompletableFuture future;
+      Statistics statistics = session.statistics(id(), HotRodMetric.METRIC_DURATION);
+      long startTimestampMs = System.currentTimeMillis();
+      statistics.incrementRequests(startTimestampMs);
+      long startTimestampNanos = System.nanoTime();
+      if (HotRodOperation.PUT.equals(operation)) {
+         future = remoteCache.putAsync(key, value);
+      } else {
+         throw new IllegalArgumentException(String.format("HotRodOperation %s not implemented", operation));
+      }
+      future.thenRun(() -> {
```

Comment:
You're missing error handling - if the request fails the test will get stuck forever. In the error handler part record a failure and call `session.stop()` as the last statement.

Reply:
Will `session.stop()` stops everything?
It is acceptable to have failures during a cache operation.
What `Statistics` method should I invoke to track errors?

Reply:
`session.stop()` will stop the current session - if you have more steps in the scenario these won't be executed. Executions of other sessions (virtual users) won't be affected. While we might think about multi-step scenarios where e.g. failure to insert the entry does not prevent other requests it's a sensible default to not invoke subsequent steps.

I shall revise the statistics to accomodate non-HTTP stuff better. For now please use `incrementResets` (this means an I/O error) and if that's some sort of TimeoutException use `incrementTimeouts`.

Reply:
Done

Reply:
Not done.
I) you're registering two handlers instead of one
II) `handle()` is invoked on success, too
III) `trackResponseError` is a no-op

Reply:
Could you please double-check again?